doc(contribute): improve advice for new contributors (#212)
leanprover-community/leanprover-community.github.io: templates/contribute/index.md

## Making a Pull Request (PR)

Once you're happy with your local changes, it's time to make a pull request.

* If you haven't already asked for write access to non-master branches of the mathlib repository,
please come to https://leanprover.zulipchat.com/, introduce yourself, and ask for this permission.

* Push your changes to a branch on the main repository, if they weren't already there.

* If you've made a lot of changes/additions, try to make many PRs containing small, self-contained pieces. This helps you get feedback as you go along, and it is much easier to review. This is especially important for new contributors.

* The title of the PR should follow our [commit conventions](https://github.com/leanprover-community/lean/blob/master/doc/commit_convention.md).

## Lifecycle of a PR

We use GitHub "labels" to manage review. (Labels can only be edited by "GitHub collaborators", which is approximately the same as "people who have asked for write access".)

On the main page for a PR, on the right-hand side,
there should be a sidebar with panels "reviewers", "assignees", "labels", etc.
Click on the "labels" header to add or remove labels from the current project.

> The most important labels are "awaiting-review" and "awaiting-author". If you label your PR with **"awaiting-review"**, someone will probably "review" it within a few days (depending on the size of the PR; smaller PRs will get quicker responses). The reviewer will probably leave comments and change the label to **"awaiting-author"**. You should address each comment, clicking the "resolve conversation" button once the problem is resolved. Ideally each problem is resolved with a new commit, but there is no hard rule here. Once all requested changes are implemented, you should change the label back to "awaiting-review" to start the process over again.

Reviewer: ```suggestion
The most important labels are "awaiting-review" and "awaiting-author". If your PR builds (has a green checkmark) and you label your PR with **"awaiting-review"**, someone will probably "review" it within a few days (depending on the size of the PR; smaller PRs will get quicker responses). The reviewer will probably leave comments and change the label to **"awaiting-author"**. You should address each comment, clicking the "resolve conversation" button once the problem is resolved. Ideally each problem is resolved with a new commit, but there is no hard rule here. Once all requested changes are implemented, you should change the label back to "awaiting-review" to start the process over again.
```

Here are some tips and tricks
to make the process of contributing as smooth as possible.

* Use [Zulip](https://leanprover.zulipchat.com/) to
   discuss your contribution before and while you are working on it.
* Adhere to the guidelines:
   - The [style guide](style.html) for contributors.
   - The explanation of [naming conventions](naming.html).
   - The [documentation guidelines](doc.html).

Once you have code that you'd like to contribute, you should open a PR.
There is a [video tutorial](https://www.youtube.com/watch?v=Bnc8w9lxe8A) walking you through the process of making a PR on YouTube.

## Working on mathlib

We use `git` to manage and version control `mathlib`.
The `master` branch is the "production" version of mathlib.
It is essential that everything in the master branch compiles without errors, and there are no `sorry`s.
To ensure this, we only commit changes to `master` that have passed CI tests, and have been approved by mathlib maintainers.

While you're working on a new contribution to `mathlib`, you should do this on a different branch.

> It's okay to do this in your own fork of the `mathlib` repository,

Author: Clarified.

or you can introduce yourself on Zulip and ask for write access to non-`master` branches of the mathlib repository.
Please include your GitHub username in your request.
It's polite to prefix the branch name with your username, so it's easier for us to clean up clutter.

Typical workflow:
* At https://github.com/leanprover-community/mathlib, click "Fork" in the top right, to make your own fork of the repository.
  Your fork is at https://github.com/USER/mathlib.

>   Alternatively, if you've asked for write access you can just use https://github.com/leanprover-community/mathlib.

Author: Clarified.

* Make local changes, e.g. using Visual Studio Code using the Lean extension.
* Commit your changes using `git commit -a`.
* If you'd like to compile everthing locally to check you didn't break anything, run
`leanproject build`. This may take a long time if you modified files low down in the import hierarchy.
It's also okay to let our central CI servers do this for you.
* In order to push your changes back to the repository on github, use
  ```
  git push
  ```
  If this complains about the remote not being configured, follow the advice in the output from `git` and run
  ```
  git push --set-upstream origin my_new_branch
  ```
* If you're working on the main `mathlib` repository rather than your own fork,
  continuous integration will automatically kick in at this point.
  You can view the output by visiting
  https://github.com/leanprover-community/mathlib/tree/my_new_branch
  (There will be a green tick on the line describing the most recent commit if everything works,
  otherwise a yellow circle if CI is still working, or a red cross if something went wrong.
  Click on the red cross to see details.)

>   You can also check CI status on the command line by installing `hub` and running `hub ci-status`.

Author: I don't think I want to over-explain here. There are lots of solutions.

It's okay to do this in your own fork of the `mathlib` repository,
or you can introduce yourself on Zulip and ask for write access to non-`master` branches of the mathlib repository.
Please include your GitHub username in your request.
It's polite to prefix the branch name with your username, so it's easier for us to clean up clutter.
(Once you're making a pull request, we'll ask you to do so from a branch of the mathlib repository,
rather than from your own fork, as CI works better this way.)

Typical workflow:
* To get started, you'll need a local copy of mathlib.
* At https://github.com/leanprover-community/mathlib, click "Fork" in the top right, to make your own fork of the repository.
  Your fork is at https://github.com/USER/mathlib.
  Alternatively, if you've asked for write access you can just use https://github.com/leanprover-community/mathlib.
* Now make a local clone of the repository.
  ```
  git clone https://github.com/USER/mathlib.git
  cd mathlib
  ```
* The steps above only need to be done once (not once for each contribution).
* Now, each time you want to work on a new change to mathlib, create a new branch:
  ```

>   git checkout -b my_new_branch   # This creates a new branch

Reviewer 2: ```suggestion
  git checkout -b my_new_branch master  # This creates a new branch departing from the master branch
```
Explicitly adding `master` might avoid confusion when someone follows these steps for a 2nd contribution, if they still have an old branch checked out. Of course, there is a risk that `master` is very old. So maybe we should talk about `git fetch` as well...

Reviewer 3: I haven't read all the new file but that part looks like it should talk about `leanproject pr my_new_branch`.

Reviewer 4: This PR has been sitting for quite a while, and looks good to me except for this comment. I don't use `leanproject` for this so I don't know for sure, is `leanproject pr my_new_branch` the same as `git checkout -b my_new_branch && git reset --hard origin/master && leanproject get-cache`?

Reviewer 3: What is does is equivalent to what you wrote + setting upstream tracking branch (so that `git push` doesn't complain).

Reviewer 2: ```suggestion
  leanproject pr my_new_branch
```

Reviewer 5: Note that `leanproject pr` also moves the local `master` reference, which is likely to be surprising if someone has made some junk commits to their local master.

Reviewer 2: Hmm, I would prefer if this instruction was as unsurprising as possible

Reviewer 3: I don't understand why Eric is complaining here. If people mess up with their master branch then they will pay for it anyway. This is orthogonal to the discussion.

Here are some other frequently-used labels:

- A **"WIP"** (= work in progress) PR still needs some foundational work (e.g. maybe it still contains `sorry`s) before getting reviewed. Post a WIP if you want to announce that you're working on something you expect to finish soon.

- A **"RFC"** (= request for comment) is a PR about a change that might be controversial or need a decision from an expert about
whether to proceed at all.

- You can add **"awaiting-CI"**, which will temporarily hide the PR on the main review queue.
  This label will be automatically removed when CI is complete.

- Consider adding the **"help wanted"** label to directly solicit contributions.

> - The **"blocked-by-other-PR"** label means that some specific other PR(s) should be resolved before addressing this one. If you add the "blocked-by-other-PR" label to your PR, please include the PR numbers of the dependencies in the PR comment so that others can see at a glance which PRs should be reviewed first.

Reviewer: ```suggestion
- The **"blocked-by-other-PR"** label means that some specific other PR(s) should be resolved before addressing this one. To add the "blocked-by-other-PR" label to your PR, include the PR numbers of the dependencies in the PR comment (following the example hidden in the comment there) so that others can see at a glance which PRs should be reviewed first. The label will be added automatically by a bot and will also be removed automatically when the other PRs have been merged.
```
